File: go/aoc25/day02/find_repeated_blocks.go
package day02

import "github.com/donmahallem/aoc/go/aoc_utils/int_util"

// Assuming intInterval is defined elsewhere like this:
// type intInterval struct {
// 	Min int
// 	Max int
// }

func findRepeatedBlocks(val intInterval) map[uint64]uint64 {
	result := make(map[uint64]uint64)
	lbound := val.Min
	ubound := val.Max

	minLen := int_util.Log10Int(lbound) + 1
	maxLen := int_util.Log10Int(ubound) + 1

	for totalLength := minLen; totalLength <= maxLen; totalLength++ {
		for blockLen := uint64(1); blockLen <= totalLength/2; blockLen++ {
			if totalLength%blockLen != 0 {
				continue
			}

			k := totalLength / blockLen
			if k < 2 {
				continue
			}

			minBlock := int_util.IntPow(10, blockLen-1)
			maxBlock := int_util.IntPow(10, blockLen) - 1

			// calc  multiplier
			numerator := int_util.IntPow(10, blockLen*k) - 1
			denominator := int_util.IntPow(10, blockLen) - 1
			mult := numerator / denominator

			for block := minBlock; block <= maxBlock; block++ {
				v := block * mult

				if v < lbound {
					continue
				}
				if v > ubound {
					// as block increases, v will only get larger
					break
				}

				// Ensure v has exactly total_length digits
				lowerLimit := int_util.IntPow(10, totalLength-1)
				upperLimit := int_util.IntPow(10, totalLength)

				if v >= lowerLimit && v < upperLimit {
					result[uint64(v)] = uint64(k)
				}
			}
		}
	}

	return result
}

File: python/aoc25/day02/find_repeated_blocks.py
import math


def _find_repeated_blocks(lbound: int, ubound: int) -> dict[int, int]:
    result: dict[int, int] = {}

    min_len = len(str(lbound))
    max_len = len(str(ubound))

    for total_length in range(min_len, max_len + 1):
        for block_len in range(1, total_length // 2 + 1):
            if total_length % block_len != 0:
                continue

            projected_block_size = total_length // block_len

            mult = (10**total_length - 1) // (10**block_len - 1)

            lower_constraint = (lbound + mult - 1) // mult
            upper_constraint = ubound // mult

            min_block_val = 10 ** (block_len - 1)
            max_block_val = 10**block_len - 1

            start_block = max(min_block_val, lower_constraint)
            end_block = min(max_block_val, upper_constraint)

            for block in range(start_block, end_block + 1):
                block_value = block * mult
                if 10 ** (total_length - 1) <= block_value < 10**total_length:
                    result[block_value] = projected_block_size

    return result
